lib/mayaUsd/nodes/proxyShapeBase.cpp

```diff
@@ -1923,6 +1965,118 @@ void MayaUsdProxyShapeBase::_OnStageEditTargetChanged(
+        UiSchemaPrependHashes.insert(hash_value(op));
+    });
+
+    if (v.IsHolding<SdfTokenListOp>()) {
+        const size_t hash = hash_value(v.UncheckedGet<SdfTokenListOp>());
+        if (UiSchemaPrependHashes.count(hash)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+// This is a stripped down copy of UsdImagingDelegate::_OnUsdObjectsChanged which is the main USD
+// notification handler where paths to refresh and paths to update are compiled for the next Hydra
+// refresh. We do not gather paths as there is no simple way to know when tho flush these maps.
+//
+// This needs to stay as quick as possible since it is stuck in the middle of the notification code
+// path.
+//
+// This is a work in progress. Some improvements might be necessary in the future. The following
+// potential issues are already visible:
+//
+//  - Changing a parameter value for the first time creates the attribute, which is a kResync
+_UsdChangeType _ClassifyUsdObjectsChanged(UsdNotice::ObjectsChanged const& notice)
+{
+    using PathRange = UsdNotice::ObjectsChanged::PathRange;
+
+    auto range = notice.GetResyncedPaths();
+    if (!range.empty()) {
+        size_t ignoredCount = 0;
+        size_t resyncCount = 0;
+        for (auto it = range.begin(); it != range.end(); ++it) {
+            if (it->IsPropertyPath()) {
+                // We have a bunch of UI properties to ignore. Especially anything that comes from
+                // UI schemas.
+                if (it->GetName().rfind("ui:", 0) == 0) {
+                    ++ignoredCount;
+                    continue;
+                }
+            }
+            for (const SdfChangeList::Entry* entry : it.base()->second) {
+                for (auto&& infoChanged : entry->infoChanged) {
+                    if (infoChanged.first == UsdTokens->apiSchemas
+                        && _IsUiSchemaPrepend(infoChanged.second.second)) {
+                        ++ignoredCount;
+                    } else {
+                        ++resyncCount;
+                    }
+                }
+            }
+        }
+
+        if (ignoredCount) {
+            return resyncCount ? _UsdChangeType::kResync : _UsdChangeType::kIgnored;
+        } else {
+            return _UsdChangeType::kResync;
+        }
+    }
+
+    auto retVal = _UsdChangeType::kIgnored;
+
+    const PathRange pathsToUpdate = notice.GetChangedInfoOnlyPaths();
+    for (PathRange::const_iterator it = pathsToUpdate.begin(); it != pathsToUpdate.end(); ++it) {
+        if (it->IsAbsoluteRootOrPrimPath()) {
+            const TfTokenVector changedFields = it.GetChangedFields();
+            if (!changedFields.empty()) {
+                retVal = _UsdChangeType::kUpdate;
+            }
+        } else if (it->IsPropertyPath()) {
+            // We have a bunch of UI properties to ignore. Especially anything that comes from UI
+            // schemas.
+            if (it->GetName().rfind("ui:", 0) == 0) {
```

Comment:
If you refactor into another file, this code seems similar to code above. A helper function might be nice to abstract detecting UI fields.